File: src/main/scala/com/github/jancajthaml/number/Shared.scala
package com.github.jancajthaml.number

private[number] object Shared {

  val AL2 = 0.301029995663981195

  val CL2 = 1.4426950408889633

  val CPI = 3.141592653589793

  val ALT = 0.693147180559945309

  val LOGE10 = 2.302585092994046

  val NIT = 3

  val N30 = 1073741824

  private var precision_current: Int = 0

  var pointer: Int = 0

  var round: Int = 0

  var precision_digits_current: Int = 0

  var precision_words: Int = 0

  var precision_digits: Int = 0

  var precision_digits_out: Int = 0

  var ellog10: Int = 0

  var mp2: Int = 0

  var mp21: Int = 0

  var nw: Int = 0

  var uu1: Array[Complex] = null

  var uu2: Array[Complex] = null

  setMaximumPrecision(300)

  def setMaximumPrecision(p: Int) {
    if (p == precision_current) return
    p = Math.min(Math.abs(p), 3600)
    precision_current = p
    pointer = 7
    round = 1
    precision_digits = precision_current
    precision_digits_out = 10000000
    ellog10 = 10 - precision_digits
    precision_words = ((precision_digits / 7.224719896) + 1).toInt
    mp2 = precision_words + 2
    mp21 = mp2 + 1
    precision_digits_current = precision_digits_out
    nw = precision_words
    val n = nw + 1
    var ti = 0.0
    var j = 0
    var i = 0
    var t1 = 0.75 * n
    val m = (CL2 * Math.log(t1) + 1.0 - 5.6843418860808015e-14).toInt
    val mq = m + 2
    val nq = (Math.pow(2, mq)).toInt
    uu1 = Array.ofDim[Complex](nq)
    uu1(0) = new Complex(mq, 0)
    var ku = 1
    var ln = 1
    j = 1
    while (j <= mq) {
      t1 = Math.PI / ln
      i = 0
      while (i <= ln - 1) {
        ti = i * t1
        uu1(i + ku) = new Complex(Math.cos(ti), Math.sin(ti))
        i += 1
      }
      ku += ln
      ln <<= 1
      j += 1
    }
    var tpn = 0.0
    var k = 0
    t1 = 0.75 * n
    uu2 = Array.ofDim[Complex](mq + nq)
    ku = mq + 1
    uu2(0) = new Complex(mq, 0)
    var mm = 0
    var nn = 0
    var mm1 = 0
    var mm2 = 0
    var nn1 = 0
    var nn2 = 0
    var iu = 0
    k = 2
    while (k <= mq - 1) {
      uu2(k - 1) = new Complex(ku, 0)
      mm = k
      nn = (Math.pow(2, mm)).toInt
      mm1 = (mm + 1) / 2
      mm2 = mm - mm1
      nn1 = (Math.pow(2, mm1)).toInt
      nn2 = (Math.pow(2, mm2)).toInt
      tpn = 2.0 * Math.PI / nn
      j = 0
      while (j < nn2) {
        i = 0
        while (i < nn1) {
          iu = ku + i + j * nn1
          t1 = tpn * i * j
          uu2(iu - 1) = new Complex(Math.cos(t1), Math.sin(t1))
          i += 1
        }
        j += 1
      }
      ku += nn
      k += 1
    }
  }

  def nint(x: Double): Double = {
    if ((x < 0)) Math.ceil(x - .5) else Math.floor(x + .5)
  }

  def fSign(a: Double, b: Double): Double = {
    (if (b >= 0) Math.abs(a) else -Math.abs(a))
  }

  def log10(i: Double): Double = (Math.log(i) / LOGE10)

  def precisionToSize(precision: Int): Int = {
    val maxnw = ((precision / 7.224719896) + 4).toInt
    if ((maxnw < 8)) 8 else maxnw
  }

}

File: src/main/scala/com/github/jancajthaml/number/Shared.java
package com.github.jancajthaml.number;

abstract class Shared {

  static final double   AL2             = 0.301029995663981195;
  static final double   CL2             = 1.4426950408889633;
  static final double   CPI             = 3.141592653589793;
  static final double   ALT             = 0.693147180559945309;
  static final double   LOGE10            = 2.302585092994046;
  static final int    NIT             = 3;
  static final int    N30             = 1073741824;
  
  private static int    precision_current     = 0;
  
  static int        pointer           = 0;
  static int        round           = 0;
  static int        precision_digits_current  = 0;
  static int        precision_words       = 0;
  static int        precision_digits      = 0;
  static int        precision_digits_out    = 0; 
  static int        ellog10           = 0; 
  static int        mp2             = 0; 
  static int        mp21            = 0;
  static int        nw              = 0; 
  static Complex[]    uu1           = null; 
  static Complex[]    uu2           = null;

  static
  { setMaximumPrecision(300); }

  public static final void setMaximumPrecision(int p)
  {
    if(p == precision_current) return;
    
    p             = Math.min(Math.abs(p), 3600);
    precision_current     = p;
    pointer         = 7;
    round           = 1;
    precision_digits      = precision_current; 
    precision_digits_out    = 10000000; 
    ellog10         = 10 - precision_digits; 
    precision_words     = (int)((precision_digits /  7.224719896) + 1);
    mp2           = precision_words + 2;
    mp21            = mp2 + 1;
    precision_digits_current  = precision_digits_out; 
    nw            = precision_words;
    int n           = nw + 1;
    double ti         = 0.0;
    int j           = 0;
    int i           = 0;
    double t1         = 0.75 * n;
    int m           = (int)(CL2 * Math.log (t1) + 1.0 - 5.6843418860808015e-14);
    int mq          = m + 2;
    int nq          = (int)(Math.pow(2, mq));
    uu1           = new Complex[nq];
    uu1[0]          = new Complex(mq,0);
    int ku          = 1;
    int ln          = 1;

    for(j = 1; j<= mq; j++)
    {
      t1 = Math.PI / ln;      
      for (i = 0; i<= ln - 1; i++)
      {
        ti      = i * t1;
        uu1[i+ku] = new Complex(Math.cos (ti), Math.sin (ti));
      }

      ku  +=  ln;
      ln  <<= 1;
    }

    double tpn  = 0.0;
    int k     = 0;
    t1      = 0.75 * n;
    uu2     = new Complex[ mq+nq ];
    ku      = mq + 1;
    uu2[0]    = new Complex(mq,0);
    int mm    = 0;
    int nn    = 0;
    int mm1   = 0;
    int mm2   = 0;
    int nn1   = 0;
    int nn2   = 0;
    int iu    = 0;

    for (k = 2; k<= mq - 1; k++)
    {
      uu2[k-1]  = new Complex(ku,0);
      mm      = k;
      nn      = (int)(Math.pow(2, mm));
      mm1     = (mm + 1) / 2;
      mm2     = mm - mm1;
      nn1     = (int)(Math.pow(2, mm1));
      nn2     = (int)(Math.pow(2, mm2));
      tpn     = 2.0 * Math.PI / nn;
      
      for(j = 0; j<nn2; j++)
      {
        for(i = 0; i<nn1; i++)
        {
          iu      = ku + i + j * nn1;
          t1      = tpn * i * j;
          uu2[iu-1] = new Complex(Math.cos (t1), Math.sin (t1));
        }
      }
      
      ku += nn;
    
    }
  }

  static final double nint(double x)
  { return (x < 0)?Math.ceil(x - .5):Math.floor(x + .5); }
  
  static final double fSign(double a, double b)
  { return (b>=0 ? Math.abs(a) : -Math.abs(a));}

  static final double log10(double val)
  { return (Math.log(val)/LOGE10); }

  static final int precisionToSize(int precision)
  {
    int maxnw = (int)((precision / 7.224719896) + 4);
    return (maxnw<8)?8:maxnw;
  }

}
